src/ripple/app/tx/impl/InvariantCheck.cpp

```diff
@@ -0,0 +1,196 @@
+            {
+            case ltACCOUNT_ROOT:
+                drops_ += (*after)[sfBalance].xrp().drops();
+                break;
+            case ltPAYCHAN:
+                if (! isDelete)
+                    drops_ += ((*after)[sfAmount] - (*after)[sfBalance]).xrp().drops();
+                break;
+            case ltESCROW:
+                if (! isDelete)
+                    drops_ += (*after)[sfAmount].xrp().drops();
+                break;
+            default:
+                break;
+            }
+        }
+    };
+
+    bool
+    finalize(STTx const& tx, TER /*tec*/, beast::Journal const& j) override
+    {
+        auto fee = tx.getFieldAmount(sfFee).xrp().drops();
+        if(-1*fee <= drops_ && drops_ <= 0)
+        {
+            return true;
+        }
+        else
+        {
+            JLOG(j.error()) << "Invariant failed: XRP net change was "
+                << drops_ << " on a fee of " << fee;
+            return false;
+        }
+    };
+
+private:
+    int64_t drops_;
+};
+
+
+/**
+ * @brief Invariant: we cannot remove an account ledger entry
+ *
+ * an account root should never be the target of a delete
+ */
+struct AccountRootsNotDeleted :
+    public InvariantCheckBase
+{
+    AccountRootsNotDeleted() : accountDeleted_(false) {};
+
+    void
+    visitEntry(
+        uint256 const& /*index*/,
+        bool isDelete,
+        std::shared_ptr <SLE const> const& before,
+        std::shared_ptr <SLE const> const& after) override
+    {
+        if (isDelete && before && before->getType() == ltACCOUNT_ROOT)
+            accountDeleted_ = true;
+    };
+
+    bool
+    finalize(STTx const& /*tx*/, TER /*tec*/, beast::Journal const& j) override
+    {
+        if (accountDeleted_)
+        {
+            JLOG(j.error()) << "Invariant failed: an account-root was deleted" ;
+            return false;
+        }
+        return true;
+    };
+
+private:
+    bool accountDeleted_;
+};
+
+/**
+ * @brief Invariant: corresponding modified ledger entries should match in type
+ *
+ */
+struct LedgerEntryTypesMatch :
+    public InvariantCheckBase
+{
+    LedgerEntryTypesMatch() : typeMismatch_(false) {};
+
+    void
+    visitEntry(
+        uint256 const& /*index*/,
+        bool /*isDelete*/,
+        std::shared_ptr <SLE const> const& before,
+        std::shared_ptr <SLE const> const& after) override
+    {
+        if (before && after && before->getType() != after->getType())
+            typeMismatch_ = true;
+    };
+
+    bool
+    finalize(STTx const& /*tx*/, TER /*tec*/, beast::Journal const& j) override
+    {
+        if (typeMismatch_)
+        {
+            JLOG(j.error()) << "Invariant failed: ledger entry type mismatch" ;
+            return false;
+        }
+        return true;
+    };
+
+private:
+    bool typeMismatch_;
+};
+
+//--
+std::array <std::unique_ptr <InvariantCheckBase>, numChecks>
+getInvariantChecks()
+{
+    return std::array <std::unique_ptr <InvariantCheckBase>, numChecks> {{
```

Comment:
Conceptually, I'm fine with either: the "Sean Parent" way is neat and very "modern C++". However, there's something to be said about simplicity. I'm curious how this actually looks like in assembly. I'll have to head over to godbolt.